### PR TITLE
Added UserLocation renderMode: native

### PR DESCRIPTION
**Pull request by mattijsf**
I added a new renderMode called `native` in an attempt to support Mapbox's built-in current location indicator on both platforms. The built in indicator some benefits compared to the normal/javascript variant such as smoother transition when the user location changes and an indicator for location accuracy.

It's a very minimal implementation atm.

### CODE REVIEW

File: android/rctmgl/src/main/java/com/mapbox/rctmgl/components/location/RCTMGLNativeUserLocation.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    @SuppressLint("MissingPermission")
+    @Override
+    public void onStyleLoaded(@NonNull Style style) {
+        Context context = getContext();
+        if (!PermissionsManager.areLocationPermissionsGranted(context)) {
+            return;
+        }
+
+        LocationComponent locationComponent = mMap.getLocationComponent();
```
Comment (mfazekas) [inline android/rctmgl/src/main/java/com/mapbox/rctmgl/components/location/RCTMGLNativeUserLocation.java]: Camera also changes options on map.getLocationComponent. 

Will that not break changing between native and non native location component?


https://github.com/react-native-mapbox-gl/maps/blob/1e886177dd0d10640628734b94a13f28e5d7ed76/android/rctmgl/src/main/java/com/mapbox/rctmgl/components/camera/RCTMGLCamera.java#L357-L381

Comment (mattijsf) [reply android/rctmgl/src/main/java/com/mapbox/rctmgl/components/location/RCTMGLNativeUserLocation.java]: Good point, I'll look into it. There must be a proper way to reuse it

**Comment by RichardLindhout** [comment]
https://docs.mapbox.com/android/maps/overview/location-component/#rendermode
Won't this conflict with these settings?

**Comment by mattijsf** [comment]
True. I also noticed it. Maybe I can rename `native` to `library` or `mapbox` to prevent confusion.

This PR is still WIP for Android. Like @mfazekas pointed out this PR breaks `LocationComponentOptions` provided by `RCTMGLCamera`. I'm still working on a solution so that `Camera` and and `UserLocation` can tweak `LocationComponentOptions`

**Comment by RichardLindhout** [comment]
Thanks, looking forward to this feature. Thanks for your work!

**Comment by RichardLindhout** [comment]
Mostly for the compass so that's why I want to have the renderMode="compass" ;-)

**Comment by lgspacil** [comment]
Is there any progress on this pull request? I am interested in accessing accurate user location and heading for the app I am building. Thanks!

**Comment by anttijva** [comment]
Very much needed feature for me also.

**Comment by ldunklee** [comment]
Also would be useful.  Thanks!

**Comment by mattijsf** [comment]
No progress yet, sorry guys. There's allot of stuff that I'm doing in the mean time and I don't have much time to prioritize this one...

It's on my list! Feel free to pick it up though, only Android needs a patch so that `LocationComponentOptions` can be gracefully modified by both Camera & UserLocation.

**Comment by RobertSasak** [comment]
As this is an optional prop, I suggest that you merge your iOS changes. Simply throw explaining message for Android. 
Longer this PR will wait here harder it will be to merge.

**Comment by RichardLindhout** [comment]
Also we can enable the user location in Android if we set
https://github.com/react-native-mapbox-gl/maps/blob/master/android/rctmgl/src/main/java/com/mapbox/rctmgl/components/camera/RCTMGLCamera.java#L65
on true

*Richard Lindhout, Developer*
*+31 643424567* <+31643424567>

*webRidge.nl*

*Beatrixlaan 2, 4694EG Scherpenisse*

KVK      56845367
BTW      NL225072464B01
*IBAN    *NL47ABNA0567375412





Op zo 6 okt. 2019 om 23:38 schreef Robert Sasak <notifications@github.com>:

**Comment by lpfunding** [comment]
Any update on this?

**Comment by RichardLindhout** [comment]
@mattijsf I'm working on it right now:
Javascript updated with support for different prop named renderType instead of renderMode
https://github.com/RichardLindhout/maps/commit/2950c5dac380fcca762f1d7d54ebda9b79e60deb

Support for compass, gps etc. in Android:
https://github.com/RichardLindhout/maps/commit/b00688e4db9d30a1ab80b6e6eaa7faa51ad6f7d2

Going into iOS now

**Comment by RichardLindhout** [comment]
Looks like iOS has no support for that, looks like it's done via the followUserLocationMode=""

**Comment by RichardLindhout** [comment]
Maybe relevant on iOS: https://github.com/mapbox/mapbox-gl-native/issues/13772

**Comment by RichardLindhout** [comment]
This PR runs fine on Android and iOS with these changes, I think we should merge it.
https://github.com/RichardLindhout/maps/blob/master/javascript/components/UserLocation.js

https://github.com/RichardLindhout/maps

**Comment by RobertSasak** [comment]
@RichardLindhout I just tested your fork on iOS and it works. Can you please explain how is your fork related to this PR?

**Comment by RichardLindhout** [comment]
@RobertSasak I merged this pull request + some changes to add compass view on Android.

+ some other things like
https://github.com/react-native-mapbox-gl/maps/pull/526

+ on Android
https://github.com/react-native-mapbox-gl/maps/issues/319

```java
        MapboxMap.CancelableCallback cameraCallback = new MapboxMap.CancelableCallback() {
            @Override
            public void onCancel() {
//                mUserTrackingState = UserTrackingState.CHANGED;
            }

            @Override
            public void onFinish() {
//                mUserTrackingState = UserTrackingState.CHANGED;
            }
        };
```

**Comment by RichardLindhout** [comment]
I need to get these back on master but more polished, and I'm working on that but I have not much spare time for that.

**Comment by RobertSasak** [comment]
I thought so. So eventually you will create a new PR, right?
Let me know if I can help with something? 
I will test Android.

**Comment by kristfal** [comment]
Closing due to lack of activity. This feature has been requested multiple times, so please reopen if/when you have time :)

**Comment by SethArchambault** [comment]
Hmm, looks like this needs to be done before we can address #437 ?

**Comment by rdgomt** [comment]
Hi @RichardLindhout! Do you have updates on this?

**Comment by RichardLindhout** [comment]
@rdgomt I will work on it when I have another customer for the app that uses this feature. I'm currently busy with some other projects so can't spend time for this. I'm sorry!

In the meantime you could use my branch which has this on iOS and Android. I know this is not ideal since we need those changes back to master. But there are some other fixes needed.

I hope Mapbox would spend time again on this library since there are quite a lot of users.

**Pull request closed without merging**